docs: Exporting Open Telemetry Data in ParadeDB (#2083)
paradedb/paradedb: docs/integrations/observability/otel.mdx

---
title: Open Telemetry
---

## Overview

> ParadeDB provides <a href="https://github.com/destrex271/postgresexporter">an unofficial OpenTelemetry (OTEL) exporter</a> that allows seamless storage of telemetry data in ParadeDB, leveraging its powerful features for data management and analysis.

Reviewer: ```suggestion
ParadeDB provides <a href="https://github.com/destrex271/postgresexporter">an unofficial OpenTelemetry (OTEL) exporter</a> that enables ingesting telemetry data into ParadeDB.
```

Supported OpenTelemetry Data:

- ✅ Logs
- ✅ Traces

> - ❌ Metrics (WIP but currently unsupported)

Reviewer: ```suggestion
- ❌ Metrics (in progress)
```

## Integration Guide

To integrate ParadeDB with your OpenTelemetry Collector, include the `postgresexporter` in your otel collector configuration file:

```yaml
exporters:
  postgresexporter:
    username: "myoteluser"
    password: "myotelpassword"
    database: "oteldata"
    logs_table_name: "otellogs"
    traces_table_name: "oteltraces"
    port: 5432

>     host: "localhost"

Author: The params are exclusive to the exporter and this is a common pattern that I observed in other thrid party exporters as well.

>     host: "localhost"

Reviewer: Got it. This should be specified, then

>     host: "localhost"

Author: okk, I'll add a separate section describing the params

>     host: "localhost"

Author: done

# -------------

service:
  # Add the postgresexporter as shown below in the service section of your collector config
  pipelines:
    logs:
      receivers: [otlp]
      exporters:
        - debug
        - postgresexporter
    traces:
      receivers: [otlp]
      exporters:
        - debug
        - postgresexporter
```

> This setup enables the export of logs and traces to ParadeDB.

Author: done

---
title: Open Telemetry
---

## Overview

> ParadeDB provides <a href="https://github.com/destrex271/postgresexporter">an unofficial OpenTelemetry (OTEL) exporter</a> that enables ingesting telemetry data into ParadeDB.
> 
> Supported OpenTelemetry Data:
> 
> - ✅ Logs
> - ✅ Traces
> - ❌ Metrics (in progress)

Reviewer: ```suggestion
ParadeDB provides an [unofficial OpenTelemetry (OTEL) exporter for Postgres](https://github.com/destrex271/postgresexporter) that enables ingesting telemetry data into ParadeDB from OTEL-compatible sources.

The following observability data types are supported:

- ✅ Logs
- ✅ Traces
- ❌ Metrics (in progress)
```

> ## Integration Guide

Reviewer: ```suggestion
## Overview
```

> To integrate ParadeDB with your OpenTelemetry Collector, include the `postgresexporter` in your otel collector configuration file:

Reviewer: ```suggestion
To ship observability data to a running ParadeDB instance, include the `postgresexporter` in your OpenTelemetry Collector configuration file as follows.
```

```yaml
exporters:
  postgresexporter:
    username: "myoteluser"
    password: "myotelpassword"
    database: "oteldata"
    logs_table_name: "otellogs"
    traces_table_name: "oteltraces"
    port: 5432
    host: "localhost"

# -------------

service:

>   # Add the postgresexporter as shown below in the service section of your collector config

Reviewer: ```suggestion
```

  pipelines:
    logs:
      receivers: [otlp]
      exporters:
        - debug
        - postgresexporter
    traces:
      receivers: [otlp]
      exporters:
        - debug
        - postgresexporter
```

<ParamField body="username" required>

>   Username for the ParadeDB instance. Can be any string.

Reviewer: ```suggestion
  Postgres username of the ParadeDB instance to which the OpenTelemetry data will be shipped.
```

</ParamField>

<ParamField body="password" required>

>   Password for the ParadeDB instance. Can be any string.

Reviewer: ```suggestion
  Postgres password of the ParadeDB instance to which the OpenTelemetry data will be shipped.
```

</ParamField>

<ParamField body="database" required>

>   Name of the database instance. Can be any valid identifier for database name.

Reviewer: ```suggestion
  Name of the Postgres database of the ParadeDB instance to which the OpenTelemetry data will be shipped.
```

</ParamField>

<ParamField body="logs_table_name" required>

>   Name of the table where OpenTelemetry (OTel) logs will be stored. Can be any
>   valid identifier for table name.

Reviewer: ```suggestion
  Name of the Postgres table within the provided database name. Must be a valid, existing table.
```

Reviewer: Do the tables need to be created beforehand? Or do they need to be available names and will be created by the exporter? Can you be more precise?

Author: nope created by the exporter, will update this

</ParamField>

<ParamField body="traces_table_name" required>

>   Name of the table where OpenTelemetry (OTel) traces will be stored. Can be any
>   valid identifier for table name.

Reviewer: ```suggestion
  Name of the Postgres table within the provided database where the OpenTelemetry traces will be stored. It must be a valid, existing table.
```

</ParamField>

<ParamField body="host" default="localhost" required>

>   TCP/IP address on which the server listens for connections from client
>   applications

Reviewer: ```suggestion
  TCP/IP address on which your ParadeDB instance listens for connections from client applications.
```

</ParamField>

<ParamField body="port" default="5432" required>

>   The TCP port the server listens on.

Reviewer: ```suggestion
  The PostgreSQL port that your ParadeDB instance is configured to listen for connections on.
```

</ParamField>

> <b>Note:</b> Ensure the specified database (e.g., otel_data) exists before using
> the exporter.

Reviewer: ```suggestion
```

> ## Structure of Logs & Traces Table

Reviewer: ```suggestion
## OpenTelemetry Tables
```

> After exporting data to ParadeDB, the following table structures will be used for storage:

Reviewer: By default, the OpenTelemetry tables will have the following schemas:

> After exporting data to ParadeDB, the following table structures will be used for storage:

Reviewer: ```suggestion
By default, the OpenTelemetry tables will be created with the following schemas:
```

> <b>1. OTEL Logs Table</b>

Reviewer: ```suggestion
### Logs
```

 TraceId            | text                           |           |          |
 SpanId             | text                           |           |          |
 TraceFlags         | smallint                       |           |          |
 SeverityText       | text                           |           |          |
 SeverityNumber     | smallint                       |           |          |
 ServiceName        | text                           |           | not null |
 Body               | text                           |           |          |
 ResourceSchemaUrl  | text                           |           |          |
 ResourceAttributes | jsonb                          |           |          |
 ScopeSchemaUrl     | text                           |           |          |
 ScopeName          | text                           |           |          |
 ScopeVersion       | text                           |           |          |
 ScopeAttributes    | jsonb                          |           |          |
 LogAttributes      | jsonb                          |           |          |
Indexes:
    "otellogs_pkey" PRIMARY KEY, btree ("ServiceName", "TimestampTime")
```

> <b>2. OTEL Traces Table</b>

Reviewer: ```suggestion
### Traces
```

```
-- Full text search for logs containing the keyword 'Processing'

SELECT "TimestampTime", "ServiceName", "SeverityText", "Body"
FROM otellogs
WHERE "Body" @@@ 'processing'
ORDER BY paradedb.score("TimestampTime") DESC
LIMIT 10;
```

<Accordion title="Expected output">
  ``` TimestampTime | ServiceName | SeverityText | Body
  ----------------------------+--------------+--------------+---------------------------
  2025-01-13 10:49:41.45282 | test-service | INFO | Processing batch 1
  2025-01-13 10:49:41.453161 | test-service | INFO | Starting order processing
  2025-01-13 10:49:42.11059 | test-service | INFO | Starting order processing
  2025-01-13 10:49:43.105943 | test-service | INFO | Starting order processing
  2025-01-13 10:49:44.113809 | test-service | INFO | Starting order processing
  2025-01-13 10:49:45.777648 | test-service | INFO | Processing batch 2
  2025-01-13 10:49:45.778265 | test-service | INFO | Starting order processing
  2025-01-13 10:49:46.680701 | test-service | INFO | Starting order processing
  2025-01-13 10:49:47.757439 | test-service | INFO | Starting order processing
  2025-01-13 10:49:48.5583 | test-service | INFO | Processing batch 3 (10 rows)
  ```

> </Accordion>

Reviewer: Let's remove this, it's documented elsewhere.

- ✅ Logs
- ✅ Traces
- ❌ Metrics (in progress)

## Overview

To ship observability data to a running ParadeDB instance, include the `postgresexporter` in your OpenTelemetry Collector configuration file as follows.

```yaml
exporters:
  postgresexporter:
    username: "myoteluser"
    password: "myotelpassword"
    database: "oteldata"
    logs_table_name: "otellogs"
    traces_table_name: "oteltraces"
    port: 5432
    host: "localhost"

# -------------

service:
  pipelines:
    logs:
      receivers: [otlp]
      exporters:
        - debug
        - postgresexporter
    traces:
      receivers: [otlp]
      exporters:
        - debug
        - postgresexporter
```

<ParamField body="username" required>
  Postgres username of the ParadeDB instance to which the OpenTelemetry data
  will be shipped.
</ParamField>

<ParamField body="password" required>
  Postgres password of the ParadeDB instance to which the OpenTelemetry data
  will be shipped.
</ParamField>

<ParamField body="database" required>
  Name of the Postgres database of the ParadeDB instance to which the
  OpenTelemetry data will be shipped.
</ParamField>

<ParamField body="logs_table_name" required>

>   Name of the Postgres table within the provided database where the
>   OpenTelemetry logs will be stored. It must be a valid table name which is then
>   created by the exporter.

Reviewer: ```suggestion
  Name of the Postgres table within the provided database where the
  OpenTelemetry logs will be stored. It must be a valid, unused table name in the provided database. The exporter will create the table.
```

</ParamField>

<ParamField body="traces_table_name" required>

>   Name of the Postgres table within the provided database where the
>   OpenTelemetry traces will be stored. It must be a valid table name which is
>   then created by the exporter.

Reviewer: ```suggestion
Name of the Postgres table within the provided database where the OpenTelemetry traces will be stored. It must be a valid, unused table name in the provided database. The exporter will create the table.
```

```
-- Full text search for logs containing the keyword 'Processing'

SELECT "TimestampTime", "ServiceName", "SeverityText", "Body"
FROM otellogs
WHERE "Body" @@@ 'processing'
ORDER BY paradedb.score("TimestampTime") DESC
LIMIT 10;
```

<Accordion title="Expected output">
  ``` TimestampTime | ServiceName | SeverityText | Body
  ----------------------------+--------------+--------------+---------------------------
  2025-01-13 10:49:41.45282 | test-service | INFO | Processing batch 1
  2025-01-13 10:49:41.453161 | test-service | INFO | Starting order processing
  2025-01-13 10:49:42.11059 | test-service | INFO | Starting order processing
  2025-01-13 10:49:43.105943 | test-service | INFO | Starting order processing
  2025-01-13 10:49:44.113809 | test-service | INFO | Starting order processing
  2025-01-13 10:49:45.777648 | test-service | INFO | Processing batch 2
  2025-01-13 10:49:45.778265 | test-service | INFO | Starting order processing
  2025-01-13 10:49:46.680701 | test-service | INFO | Starting order processing
  2025-01-13 10:49:47.757439 | test-service | INFO | Starting order processing
  2025-01-13 10:49:48.5583 | test-service | INFO | Processing batch 3 (10 rows)
  ```

> </Accordion>

Reviewer: If you want to keep this, actually, I like it. I think we should convert it to an accordion so it's not clogging the documentation page

Author: cool will do that

Author: done
